ENH - Update translations workflows (#1959)
pydata/pydata-sphinx-theme: docs/community/topics/i18n.rst

-----------------------------------------------

All natural language text in the theme's components and layouts must be marked as translatable so that it can be extracted and translated.
For example, if you add a button with the text "Next page", you will need to mark this text as translatable.

To do so, you can use the Jinja2 ``trans`` block and/or a ``_()`` function to mark text as translatable in corresponding ``HTML`` templates (located in the ``src/pydata_sphinx_theme/theme/`` directory).
`Please refer to the Jinja2 documentation <https://jinja.palletsprojects.com/en/2.11.x/templates/#i18n>`__.
Remember to `manually escape variables <https://jinja.palletsprojects.com/en/2.11.x/templates/#working-with-manual-escaping>`__ if needed.

For example, to mark the text "Next page" as translatable, you would write:

.. code-block:: jinja
   
   <button type="button">
      {{- _("Next page") -}}
   </button>

Any text that is marked in this way will be discoverable by ``pybabel`` and used to generate the ``PO`` files.
Once you've marked the text as translatable, complete the steps outlined in :ref:`changing-natural-language-text`.

.. _changing-natural-language-text:

> Adding or changing natural language text

Reviewer: the prior section is called "marking natural language text as translatable"; it seems like this section should be called

```suggestion
Updating the translation catalogs
```

...or something like that

Author: 👍🏽 thanks I was struggling with some of the wording and content, so will be making more changes.

-----------------------------------------

> If you add or change natural language text in the theme, you must update the message catalogs to include the new or updated text. Follow these steps:

Reviewer: [nitpick]

```suggestion
When you add or change natural language text in the theme, you must update the message catalogs to include the new or updated text. Follow these steps:
```

#. Edit the natural language text and ensure it is :ref:`marked as translatable <adding-translatable-text>`.

#. Generate/update the message catalog template (``POT`` file):

   .. code-block:: bash
      
      tox run -e i18n-extract

#. Update the message catalogs (``PO`` files) for the existing locales:

   .. code-block:: bash

>       tox run -e i18n-update

Reviewer: since the workflow is 

```bash
tox run -e i18n-extract
tox run -e i18n-update
```

...with apparently no steps in between, I find myself wondering why these are two separate steps (from the contributor perspective).  Why not a single tox command to do both things?

Author: 🤔 all the sources and projects I surveyed to get an idea of the workflow treat these as a separate step... will have another thought and it might make sense to merge

Reviewer: IIUC, I guess it makes sense sometimes to do `update` without doing `extract` (i.e., if only non-translatable stuff changed, and you just want to update the line numbers)?  But even so it would still be possible to have a single recipe to do both, and a sub-recipe that only did the update part.

Reviewer 2: That makes total sense to me based on the [workflow](https://github.com/pydata/pydata-sphinx-theme/pull/1959#pullrequestreview-2266707690) I followed.

This will update the translation files with new information about the position and text of the language you have modified.

If you *only* change non-translatable text (like HTML markup), the ``extract`` and ``update`` commands will only update the
positions (line numbers) of the translatable strings.
Updating positions is optional - the line numbers are to inform the human translator, not to perform the translation.

If you change translatable strings, the ``extract`` command will extract the new or updated strings to the ``POT`` file,
and the ``update`` command will try to perform a fuzzy match between the new or updated strings and existing translations in the ``PO`` files.
If there is a fuzzy match, a comment like ``#, fuzzy`` is added before the matched entry,
this means that the translation needs to be manually reviewed and possibly updated.
If after reviewing the translation you decide to keep the existing translation, you can remove the ``#, fuzzy`` comment from the entry.

> Otherwise, it will add a new translation entry.

Reviewer: the scope of this "otherwise" is a bit unclear to me. Does it mean "if there is a non-fuzzy match" (i.e., a perfect match)?  Or does it mean "if there is no match"?  I think it's the latter.  It might help to switch the order of information, like:

> If no match is found, a new translation entry will be automatically added.
> If a match *is* found, a comment is added to the matched entry...

Author: I mean if there is no fuzzy match - noting for rewording here

.. _adding-new-language:

Adding a new language
----------------------

> The list of currently supported languages is available in the ``src/pydata_sphinx_theme/locale`` directory.

Reviewer: ```suggestion
The list of languages with existing (possibly incomplete) translations is available in the ``src/pydata_sphinx_theme/locale`` directory.
```

To add a new language, follow these steps:

#. Identify the `ISO 639-1 code <https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes>`__ for the new language.

#. Make a copy of the ``POT`` file for this new language:

   .. code-block:: bash
      
      # for example, to add Quechua (ISO 639-1 code: qu)

>       tox -e i18n-new-locale -- qu

Reviewer: ```suggestion
      tox -e i18n-new-locale -- qu

#. Start translating the theme's text into the newly added language (see :ref:`translating-the-theme`).
```

I know it's the immediate next section, but I still think it's useful to be explicit that "adding a new language" involves some easy/automated steps but also a lot of skilled human work too

Internationalization
====================

We use `Gettext <https://docs.python.org/3/library/gettext.html>`__ message catalogs for the theme's internationalization (I18N) and localization (L10N) and `pybabel <https://babel.pocoo.org/en/latest/messages.html>`__ to work with them.

User-facing strings in the theme should be internationalized, so that they can be translated into other languages.
This section covers how to mark text as translatable and how to translate the theme.

..Note

>    Internationalization (or i18n) is the process of marking strings for translation so that the strings can be extracted from the source code and given to translators. Localization (l10n) is the process of translating the marked strings into different languages.

Author: Funny enough localisation is also used to refer to what you mentioned: dates, currency, etc. 
Depending on the translation context I suppose it will or not include items like this.

>    Internationalization (or i18n) is the process of marking strings for translation so that the strings can be extracted from the source code and given to translators. Localization (l10n) is the process of translating the marked strings into different languages.

Reviewer 3: I remember it because localisation is not only about language, but other things. Like fr_fr, and fr_ca will have different words and units, even if both are french. 

It's also coherent with the usage of the term [LOCALE](https://en.wikipedia.org/wiki/Locale_(computer_software)), 

> In [computing](https://en.wikipedia.org/wiki/Computing), a locale is a set of [parameters](https://en.wikipedia.org/wiki/Parameter) that defines the user's language, region and any special variant preferences that the user wants to see in their [user interface](https://en.wikipedia.org/wiki/User_interface). Usually a locale identifier consists of at least a language code and a country/region code.

But yes, that's the terms of i18n and l10n (though I find the 18 and 10 horrible, and I think it should have been fully spelled, but not my decision).

>    Internationalization (or i18n) is the process of marking strings for translation so that the strings can be extracted from the source code and given to translators. Localization (l10n) is the process of translating the marked strings into different languages.

Reviewer 2: I tend to agree. I'm not a big fan of numeronyms. But I do make an exception for i18n because 18 starts to be quite a lot. I cannot seem to type internationalization without making a mistake, and if English were my second language I bet it would also be a hard word to say, whereas localization doesn't trip me up.

The general process for internationalizing and localizing the theme is as follows:

#. Mark natural language text as translatable.
#. Extract translatable text to a message catalog template ``POT`` (``PO`` template file).
#. Make a copy of the ``POT`` file for each language (locale) you want to translate (or update existing templates).

> #. Compile the message catalogs to binary ``MO`` files.

Author: Basically this is what is used at runtime.

> #. Compile the message catalogs to binary ``MO`` files.

Reviewer 3: Translation are more generally used than for docs, it's also used for software that run locally, it is therefor important to be small and fast to read and parse. Here we happen render docs statically but it's not relevant for gettext. So here runtime is when you rendre the docs.

> #. Compile the message catalogs to binary ``MO`` files.

Reviewer 2: I see. So the MO files could be used dynamically, or at runtime, by, say, a desktop application after it loads but before it has rendered the UI. And this same desktop application might allow you to switch languages while it's running, which is when having a compact MO file that's quick to read and parse comes in handy.

      {{- _("Next page") -}}
   </button>

Any text that is marked in this way will be discoverable by ``pybabel`` and used to generate the ``PO`` files.
Once you've marked the text as translatable, complete the steps outlined in :ref:`changing-natural-language-text`.

.. _changing-natural-language-text:

Updating the translation catalogs
-----------------------------------------

When you add or change natural language text in the theme, you must update the message catalogs to include the new or updated text. Follow these steps:

#. Edit the natural language text and ensure it is :ref:`marked as translatable <adding-translatable-text>`.

#. Generate/update the message catalog template (``POT`` file):

   .. code-block:: bash

      tox run -e i18n-extract

#. Update the message catalogs (``PO`` files) for the existing locales:

   .. code-block:: bash

      tox run -e i18n-update

This will update the translation files with new information about the position and text of the language you have modified.

If you *only* change non-translatable text (like HTML markup), the ``extract`` and ``update`` commands will only update the
positions (line numbers) of the translatable strings.

> Updating positions is optional - the line numbers are to inform the human translator, not to perform the translation.

Author: Ideally yes, but if that does not happen every time the fuzzy finding capabilities in pybabel help finding the nearest match

#. Generate a ``PO`` file based on ``POT`` file for this new language:

   .. code-block:: bash

      # for example, to add Quechua (ISO 639-1 code: qu)
      tox -e i18n-new-locale -- qu

#. Update and compile the localization files as described in the :ref:`updating-localization-files` and 
   :ref:`compiling-localization-files` sections. Then commit the changes.
#. Localize the theme's into the newly added language (see :ref:`localizing-the-theme`).

.. _localizing-the-theme:

Localizing the theme
---------------------

We manage localizations on the `PyData Sphinx Theme project on Transifex <https://explore.transifex.com/12rambau/pydata-sphinx-theme/>`__.

To contribute localization, follow these steps:

#. Sign up for a `Transifex account <https://www.transifex.com/signup/>`__.
#. Join the `PyData Sphinx Theme project <https://explore.transifex.com/12rambau/pydata-sphinx-theme/>`__.
#. Select the language you want to localize. If the language you are looking for is not listed, you can `open an issue on GitHub to request it <https://github.com/pydata/pydata-sphinx-theme/issues>`__. Then you can open a pull request 

>    to add the new language following the steps outlined in :ref:`adding-new-language`.

Author: Had to revert as the previous one was making the docs build fail

>    to add the new language following the steps outlined in :ref:`adding-a-new-language`.

Reviewer: I think for autosection-based cross-refs you leave in the spaces:

```suggestion
   to add the new language following the steps outlined in :ref:`adding a new language`.
```

Full sentences and clauses must always be a single translatable string.
Otherwise, you can get ``next page`` translated as ``suivant page`` instead of as ``page suivante``, etc.

Dealing with variables and markup in translations
`````````````````````````````````````````````````````````````

If a variable (like the ``edit_page_provider_name`` theme option) is used as part of a phrase, it must be included within the translatable string.

> Otherwise, the word order in other languages can be incorrect. In a Jinja template, simply surround the translatable string with ``{% trans variable=variable %}`` and ``{% endtrans %}}``. For example:

Reviewer: ```suggestion
Otherwise, the word order in other languages can be incorrect. In a Jinja template, surround the translatable string with ``{% trans variable=variable %}`` and ``{% endtrans %}``. For example:
```

.. code-block:: jinja

   {% trans provider=provider %}Edit on {{ provider }}{% endtrans %}
   
The translatable string is extracted as the Python format string ``Edit on %(provider)s``. This is so that the same translatable string can be used in both Python code and Jinja templates. It is the translator's responsibility to use ``%(provider)s`` verbatim in the translation.

If a non-translatable word or token (like HTML markup) is used as part of a phrase, it must also be included within the translatable string. For example:

.. code-block:: jinja

   {% trans theme_version=theme_version|e %}
   Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> {{ theme_version }}.
   {% endtrans %}

It is the translator's responsibility to use the HTML markup verbatim in the translation.

Translation files

> ------------------

Reviewer: some of the info in this section (e.g., "there are 3 types of files, here's what they are") would have been useful context at the beginning of the page...  but some of it (e.g., details about `#:` vs `#,`) might be too "in the weeds" to come right at the beginning. Consider splitting up?

Author: This is a section I struggled with -- content and position -- so perhaps splitting would be good. 

I think the _too detailed_ bits can go under the "Tips" section...

Full sentences and clauses must always be a single translatable string.
Otherwise, you can get ``next page`` translated as ``suivant page`` instead of as ``page suivante``, etc.

> Dealing with variables and markup in translations
> `````````````````````````````````````````````````````````````

Reviewer 3: ```suggestion
Dealing with variables and markup in translations
`````````````````````````````````````````````````
```

OCD.

Reviewer 3: Rahaaa. github does not understand, I reduced the number of backitcks to be the same length as the title.
